### ISSUE TITLE
FR: add option to control charging in Kobo devices to protect battery

### ISSUE THREAD

**Issue by cgm999** [koreader/koreader#12278]
My device is Kobo Sage and I noticed that if write 0 to /sys/class/power_supply/battery/charging it will .. stop the charging.
Not sure about other devices, however it would be nice if there would be an option in Koreader to control the bottom and top levels of the charging to protect the battery. The easiest logic to me it would be something like:

- stop charging if level is above top_level and only renable charging when level drops below bottom_level
- provide some option to charge one time to a specified level

**Comment by hius07** [comment]
We have battery level watch in Settings - Device - Device status alerts, but it shows alerts (warnings) only, no real charging control.

**Comment by NiLuJe** [comment]
That's because we can't actually control it ;).

*Some* devices do offer *very* minimal userland PMIC control, but that's the exception rather than the norm, and it's usually a massively buggy mess (case in point: the sunxi family, which I think @zwim spent a long time trying to cobble into submission, to no avail).

TL;DR: can't fix, won't fix, please god no stop the hgoreroror the insanity ntx kernels arraagahagzhahagah.

**Comment by NiLuJe** [comment]
For context on how miraculous it is that basic shit actually (sometimes) work as far as PMICs and NTX boards are concerned: the latest lineup of MTK boards will simply *crash* if you try to enter suspend/standby while *charging* (lol).

So, asking them to do more fancy stuff? Hahahahah, hell no.

**Comment by Frenzie** [comment]
> the latest lineup of MTK boards will simply _crash_ if you try to enter suspend/standby while _charging_ (lol).

And there I thought it was mildly annoying that some (most/all?) refuse to in that condition.

**Comment by NiLuJe** [comment]
> And there I thought it was mildly annoying that some (most/all?) refuse to in that condition.

They managed to do it right in the latest rounds of NXP devices (Mk. 7+ for sure, possibly as early as Mk. 6), they'll even properly tear down usb gagdets beforehand (e.g., usbnet).

**Comment by cgm999** [comment]
Can an option by added to koreader to disable auto-suspend if charging is on ? This will make life easier for a custom script that messes with charging on/off

**Comment by Frenzie** [comment]
That's the behavior as is though?

**Comment by cgm999** [comment]
I do not think that is true, I left a sleep running in background with reboot after .. and it does not reboot. Something like this:
sh -c ' ( sleep 30 ; reboot ) &'

**Comment by Frenzie** [comment]
To be clear, that means charging, not charged and plugged in.

**Comment by cgm999** [comment]
ah darn, I said autosuspend, I ment autostandby .. and yes my kobo sage is carging and still goes to standby

**Comment by Frenzie** [comment]
If that's the case then it's a Sage-specific issue. It doesn't happen on the H2O v1 or the Libra 2.

Unless of course it's charged as opposed to charging.

**Comment by cgm999** [comment]
is charging. I see it on usb metter.. will try to see if I can figure why is not working on my Sage

**Comment by Frenzie** [comment]
A USB meter doesn't matter, only `/sys/class/power_supply/battery/status` or depending on the device one of its friends `/sys/class/power_supply/bd71827_bat/status` or `/sys/class/power_supply/mc13892_bat/status` does.

https://github.com/koreader/koreader/blob/2d2921032d176e944d49412a1ff157d4726958db/frontend/device/kobo/device.lua#L733-L740

**Comment by NiLuJe** [comment]
You *would* see a drop on a meter (... on properly engineered boards, at least), but that's indeed more of an observation; we base our actual decision on what the PMIC says, as @Frenzie just mentioned.

There is a framework in place for something like what you asked, mainly because most boards will horribly implode when you try it, so we have to keep a blacklist.

https://github.com/koreader/koreader/blob/2d2921032d176e944d49412a1ff157d4726958db/frontend/device/kobo/device.lua#L956-L961

TL;DR: Switch the  `canPowerSaveWhileCharging` devcap to `no` in a late patch (e.g., like the no-hw-invert one in the wiki).

**Comment by cgm999** [comment]
Yep thats it..  canPowerSaveWhileCharging set to no. Yep 10x , I will see , my plan is to use a dumb .sh script that when is seeing
battery over 80% to stop charging with sleep 60 between checks.
I will report here (for whoever "likes pain ") how it works
